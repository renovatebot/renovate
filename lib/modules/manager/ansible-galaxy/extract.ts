import { logger } from '../../../logger';
import { newlineRegex, regEx } from '../../../util/regex';
import type { PackageDependency, PackageFileContent } from '../types';
import { extractCollections } from './collections';
import { extractCollectionsMetaDataFile } from './collections-metadata';
import { extractRoles } from './roles';

export function getSliceEndNumber(
  start: number,
  numberOfLines: number,
  ...blocks: number[]
): number {
  if (start < 0 || start > numberOfLines - 1) {
    return -1;
  }
  let nearestEnd = numberOfLines;
  for (const blocksKey of blocks) {
    if (start < blocksKey && blocksKey < nearestEnd) {
      nearestEnd = blocksKey;
    }
  }
  return nearestEnd;
}

export function extractPackageFile(
  content: string,
  packageFile: string
): PackageFileContent | null {
  logger.trace(`ansible-galaxy.extractPackageFile(${packageFile})`);
  const galaxyFileNameRegEx = regEx(/galaxy\.ya?ml$/);
  const deps: PackageDependency[] = [];
  const lines = content.split(newlineRegex);

  try {
    // if this is a galaxy.yml file we have to interpret the dependencies differently
    if (galaxyFileNameRegEx.exec(packageFile)) {
      const galaxyDeps = extractCollectionsMetaDataFile(lines);
      deps.push(...galaxyDeps);
    } else {
      // interpret requirements file
      // check if new or old format is used and save start lines for collection and roles.
      const positions = {
        collections: -1,
        roles: -1,
      };
      // find role and collection block
      lines.forEach((line, index) => {
        if (regEx(/^collections:/).exec(line)) {
          positions.collections = index;
        }
        if (regEx(/^roles:/).exec(line)) {
          positions.roles = index;
        }
      });
      if (positions.collections >= 0 || positions.roles >= 0) {
        // using new format
        const collectionLines = lines.slice(
          positions.collections,
          getSliceEndNumber(
            positions.collections,
            lines.length,
            positions.roles
          )
        );
        const collectionDeps = extractCollections(collectionLines);
        deps.push(...collectionDeps);

        const roleLines = lines.slice(
          positions.roles,
          getSliceEndNumber(
            positions.roles,
            lines.length,
            positions.collections
          )
        );
        const roleDeps = extractRoles(roleLines);
        deps.push(...roleDeps);
      } else {
        // use old format which only has only roles
        const galaxyDeps = extractRoles(lines);
        deps.push(...galaxyDeps);
      }
    }

    if (!deps.length) {
      return null;
    }
    return { deps };
  } catch (err) /* istanbul ignore next */ {
    logger.debug({ err, packageFile }, 'Error extracting ansible-galaxy deps');
    return null;
  }
}
