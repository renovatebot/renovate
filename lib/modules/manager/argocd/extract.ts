import is from '@sindresorhus/is';
import { loadAll } from 'js-yaml';
import { logger } from '../../../logger';
import { coerceArray } from '../../../util/array';
import { trimTrailingSlash } from '../../../util/url';
import { DockerDatasource } from '../../datasource/docker';
import { GitTagsDatasource } from '../../datasource/git-tags';
import { HelmDatasource } from '../../datasource/helm';
import type {
  ExtractConfig,
  PackageDependency,
  PackageFileContent,
} from '../types';
import type {
  ApplicationDefinition,
  ApplicationSource,
  ApplicationSpec,
} from './types';
import { fileTestRegex } from './util';

export function extractPackageFile(
  content: string,
  packageFile: string,
  _config?: ExtractConfig
): PackageFileContent | null {
  // check for argo reference. API version for the kind attribute is used
  if (fileTestRegex.test(content) === false) {
    return null;
  }

  let definitions: ApplicationDefinition[];
  try {
    definitions = loadAll(content) as ApplicationDefinition[];
  } catch (err) {
    logger.debug({ err, packageFile }, 'Failed to parse ArgoCD definition.');
    return null;
  }

  const deps = definitions.filter(is.plainObject).flatMap(processAppSpec);

  return deps.length ? { deps } : null;
}

function processSource(source: ApplicationSource): PackageDependency | null {
  if (
    !source ||
    !is.nonEmptyString(source.repoURL) ||
    !is.nonEmptyString(source.targetRevision)
  ) {
    return null;
  }

  // a chart variable is defined this is helm declaration
  if (source.chart) {
    // assume OCI helm chart if repoURL doesn't contain explicit protocol
    if (
      source.repoURL.startsWith('oci://') ||
      !source.repoURL.includes('://')
    ) {
      let registryURL = source.repoURL.replace('oci://', '');
      registryURL = trimTrailingSlash(registryURL);

      return {
        depName: `${registryURL}/${source.chart}`,
        currentValue: source.targetRevision,
        datasource: DockerDatasource.id,
      };
    }

    return {
      depName: source.chart,
      registryUrls: [source.repoURL],
      currentValue: source.targetRevision,
      datasource: HelmDatasource.id,
    };
  }

  return {
    depName: source.repoURL,
    currentValue: source.targetRevision,
    datasource: GitTagsDatasource.id,
  };
}

function processAppSpec(
  definition: ApplicationDefinition
): PackageDependency[] {
  const spec: ApplicationSpec | null | undefined =
    definition.kind === 'Application'
      ? definition?.spec
      : definition?.spec?.template?.spec;

  if (is.nullOrUndefined(spec)) {
    return [];
  }

  const deps: (PackageDependency | null)[] = [];

  if (is.nonEmptyObject(spec.source)) {
    deps.push(processSource(spec.source));
  }

  for (const source of coerceArray(spec.sources)) {
    deps.push(processSource(source));
  }

  return deps.filter(is.truthy);
}
