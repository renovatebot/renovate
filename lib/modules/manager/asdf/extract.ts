import is from '@sindresorhus/is';
import { logger } from '../../../logger';
import { isSkipComment } from '../../../util/ignore';
import { regEx } from '../../../util/regex';
import type { PackageDependency, PackageFileContent } from '../types';
import { upgradeableTooling } from './upgradeable-tooling';

export function extractPackageFile(content: string): PackageFileContent | null {
  logger.trace(`asdf.extractPackageFile()`);

  const regex = regEx(
    /^(?<toolName>([\w_-]+)) (?<version>[^\s#]+)(?: [^\s#]+)* *(?: #(?<comment>.*))?$/gm
  );

  const deps: PackageDependency[] = [];

  for (const groups of [...content.matchAll(regex)]
    .map((m) => m.groups)
    .filter(is.truthy)) {
    const depName = groups.toolName.trim();
    const version = groups.version.trim();

    const toolConfig = upgradeableTooling[depName];
    const toolDefinition = toolConfig
      ? typeof toolConfig.config === 'function'
        ? toolConfig.config(version)
        : toolConfig.config
      : undefined;

    if (toolDefinition) {
      const dep: PackageDependency = {
        currentValue: version,
        depName,
        ...toolDefinition,
      };
      if (isSkipComment((groups.comment ?? '').trim())) {
        dep.skipReason = 'ignored';
      }

      deps.push(dep);
    } else {
      const dep: PackageDependency = {
        depName,
        skipReason: 'unsupported-datasource',
      };

      deps.push(dep);
    }
  }

  return deps.length ? { deps } : null;
}
