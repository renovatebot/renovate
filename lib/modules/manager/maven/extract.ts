import is from '@sindresorhus/is';
import upath from 'upath';
import { XmlDocument, XmlElement } from 'xmldoc';
import { logger } from '../../../logger';
import { readLocalFile } from '../../../util/fs';
import { regEx } from '../../../util/regex';
import { MavenDatasource } from '../../datasource/maven';
import { MAVEN_REPO } from '../../datasource/maven/common';
import type { ExtractConfig, PackageDependency, PackageFile } from '../types';
import type { MavenProp } from './types';

function parsePom(raw: string, packageFile: string): XmlDocument | null {
  let project: XmlDocument;
  try {
    project = new XmlDocument(raw);
  } catch (err) {
    logger.debug({ packageFile }, `Failed to parse as XML`);
    return null;
  }
  const { name, attr, children } = project;
  if (name !== 'project') {
    return null;
  }
  if (attr.xmlns === 'http://maven.apache.org/POM/4.0.0') {
    return project;
  }
  if (
    is.nonEmptyArray(children) &&
    children.some((c: any) => c.name === 'modelVersion' && c.val === '4.0.0')
  ) {
    return project;
  }
  return null;
}

function containsPlaceholder(str: string | null | undefined): boolean {
  return !!str && regEx(/\${[^}]*?}/).test(str);
}

function depFromNode(
  node: XmlElement,
  underBuildSettingsElement = false
): PackageDependency | null {
  if (!('valueWithPath' in node)) {
    return null;
  }
  let groupId = node.valueWithPath('groupId')?.trim();
  const artifactId = node.valueWithPath('artifactId')?.trim();
  const currentValue = node.valueWithPath('version')?.trim();
  let depType: string | undefined;

  if (!groupId && node.name === 'plugin') {
    groupId = 'org.apache.maven.plugins';
  }

  if (groupId && artifactId && currentValue) {
    const depName = `${groupId}:${artifactId}`;
    const versionNode = node.descendantWithPath('version')!;
    const fileReplacePosition = versionNode.position;
    const datasource = MavenDatasource.id;
    const registryUrls = [MAVEN_REPO];
    const result: PackageDependency = {
      datasource,
      depName,
      currentValue,
      fileReplacePosition,
      registryUrls,
    };

    switch (node.name) {
      case 'plugin':
      case 'extension':
        depType = 'build';
        break;
      case 'parent':
        depType = 'parent';
        break;
      case 'dependency':
        if (underBuildSettingsElement) {
          depType = 'build';
        } else if (node.valueWithPath('optional')?.trim() === 'true') {
          depType = 'optional';
        } else {
          depType = node.valueWithPath('scope')?.trim() ?? 'compile'; // maven default scope is compile
        }
        break;
    }

    if (depType) {
      result.depType = depType;
    }

    return result;
  }
  return null;
}

function deepExtract(
  node: XmlElement,
  result: PackageDependency[] = [],
  isRoot = true,
  underBuildSettingsElement = false
): PackageDependency[] {
  const dep = depFromNode(node, underBuildSettingsElement);
  if (dep && !isRoot) {
    result.push(dep);
  }
  if (node.children) {
    for (const child of node.children) {
      deepExtract(
        child as XmlElement,
        result,
        false,
        node.name === 'build' ||
          node.name === 'reporting' ||
          underBuildSettingsElement
      );
    }
  }
  return result;
}

function applyProps(
  dep: PackageDependency<Record<string, any>>,
  depPackageFile: string,
  props: MavenProp
): PackageDependency<Record<string, any>> {
  let result = dep;
  let anyChange = false;
  const alreadySeenProps: Set<string> = new Set();

  do {
    const [returnedResult, returnedAnyChange, fatal] = applyPropsInternal(
      result,
      depPackageFile,
      props,
      alreadySeenProps
    );
    if (fatal) {
      dep.skipReason = 'recursive-placeholder';
      return dep;
    }
    result = returnedResult;
    anyChange = returnedAnyChange;
  } while (anyChange);

  if (containsPlaceholder(result.depName)) {
    result.skipReason = 'name-placeholder';
  } else if (containsPlaceholder(result.currentValue)) {
    result.skipReason = 'version-placeholder';
  }

  return result;
}

function applyPropsInternal(
  dep: PackageDependency<Record<string, any>>,
  depPackageFile: string,
  props: MavenProp,
  previouslySeenProps: Set<string>
): [PackageDependency<Record<string, any>>, boolean, boolean] {
  let anyChange = false;
  let fatal = false;

  const seenProps: Set<string> = new Set();

  const replaceAll = (str: string): string =>
    str.replace(regEx(/\${[^}]*?}/g), (substr) => {
      const propKey = substr.slice(2, -1).trim();
      // TODO: wrong types here, props is already `MavenProp`
      const propValue = (props as any)[propKey] as MavenProp;
      if (propValue) {
        anyChange = true;
        if (previouslySeenProps.has(propKey)) {
          fatal = true;
        } else {
          seenProps.add(propKey);
        }
        return propValue.val;
      }
      return substr;
    });

  const depName = replaceAll(dep.depName!);
  const registryUrls = dep.registryUrls!.map((url) => replaceAll(url));

  let fileReplacePosition = dep.fileReplacePosition;
  let propSource = dep.propSource;
  let groupName: string | null = null;
  const currentValue = dep.currentValue!.replace(
    regEx(/^\${[^}]*?}$/),
    (substr) => {
      const propKey = substr.slice(2, -1).trim();
      // TODO: wrong types here, props is already `MavenProp`
      const propValue = (props as any)[propKey] as MavenProp;
      if (propValue) {
        if (!groupName) {
          groupName = propKey;
        }
        fileReplacePosition = propValue.fileReplacePosition;
        propSource = propValue.packageFile ?? undefined;
        anyChange = true;
        if (previouslySeenProps.has(propKey)) {
          fatal = true;
        } else {
          seenProps.add(propKey);
        }
        return propValue.val;
      }
      return substr;
    }
  );

  const result: PackageDependency = {
    ...dep,
    depName,
    registryUrls,
    fileReplacePosition,
    propSource,
    currentValue,
  };

  if (groupName) {
    result.groupName = groupName;
  }

  if (propSource && depPackageFile !== propSource) {
    result.editFile = propSource;
  }

  for (const prop of seenProps) {
    previouslySeenProps.add(prop);
  }
  return [result, anyChange, fatal];
}

function resolveParentFile(packageFile: string, parentPath: string): string {
  let parentFile = 'pom.xml';
  let parentDir = parentPath;
  const parentBasename = upath.basename(parentPath);
  if (parentBasename === 'pom.xml' || parentBasename.endsWith('.pom.xml')) {
    parentFile = parentBasename;
    parentDir = upath.dirname(parentPath);
  }
  const dir = upath.dirname(packageFile);
  return upath.normalize(upath.join(dir, parentDir, parentFile));
}

interface MavenInterimPackageFile extends PackageFile {
  mavenProps?: Record<string, any>;
  parent?: string;
}

export function extractPackage(
  rawContent: string,
  packageFile: string
): PackageFile | null {
  if (!rawContent) {
    return null;
  }

  const project = parsePom(rawContent, packageFile);
  if (!project) {
    return null;
  }

  const result: MavenInterimPackageFile = {
    datasource: MavenDatasource.id,
    packageFile,
    deps: [],
  };

  result.deps = deepExtract(project);

  const propsNode = project.childNamed('properties');
  const props: Record<string, MavenProp> = {};
  if (propsNode?.children) {
    for (const propNode of propsNode.children as XmlElement[]) {
      const key = propNode.name;
      const val = propNode?.val?.trim();
      if (key && val) {
        const fileReplacePosition = propNode.position;
        props[key] = { val, fileReplacePosition, packageFile };
      }
    }
  }
  result.mavenProps = props;

  const repositories = project.childNamed('repositories');
  if (repositories?.children) {
    const repoUrls: string[] = [];
    for (const repo of repositories.childrenNamed('repository')) {
      const repoUrl = repo.valueWithPath('url')?.trim();
      if (repoUrl) {
        repoUrls.push(repoUrl);
      }
    }
    result.deps.forEach((dep) => {
      if (is.array(dep.registryUrls)) {
        repoUrls.forEach((url) => dep.registryUrls!.push(url));
      }
    });
  }

  if (packageFile && project.childNamed('parent')) {
    const parentPath =
      project.valueWithPath('parent.relativePath')?.trim() ?? '../pom.xml';
    result.parent = resolveParentFile(packageFile, parentPath);
  }

  if (project.childNamed('version')) {
    result.packageFileVersion = project.valueWithPath('version')!.trim();
  }

  return result;
}

export function extractRegistries(rawContent: string): string[] {
  if (!rawContent) {
    return [];
  }

  const settings = parseSettings(rawContent);
  if (!settings) {
    return [];
  }

  const urls: string[] = [];

  const mirrorUrls = parseUrls(settings, 'mirrors');
  urls.push(...mirrorUrls);

  settings.childNamed('profiles')?.eachChild((profile) => {
    const repositoryUrls = parseUrls(profile, 'repositories');
    urls.push(...repositoryUrls);
  });

  // filter out duplicates
  return [...new Set(urls)];
}

function parseUrls(xmlNode: XmlElement, path: string): string[] {
  const children = xmlNode.descendantWithPath(path);
  const urls: string[] = [];
  if (children?.children) {
    children.eachChild((child) => {
      const url = child.valueWithPath('url');
      if (url) {
        urls.push(url);
      }
    });
  }
  return urls;
}

export function parseSettings(raw: string): XmlDocument | null {
  let settings: XmlDocument;
  try {
    settings = new XmlDocument(raw);
  } catch (e) {
    return null;
  }
  const { name, attr } = settings;
  if (name !== 'settings') {
    return null;
  }
  if (attr.xmlns === 'http://maven.apache.org/SETTINGS/1.0.0') {
    return settings;
  }
  return null;
}

export function resolveParents(packages: PackageFile[]): PackageFile[] {
  const packageFileNames: string[] = [];
  const extractedPackages: Record<string, MavenInterimPackageFile> = {};
  const extractedDeps: Record<string, PackageDependency[]> = {};
  const extractedProps: Record<string, MavenProp> = {};
  const registryUrls: Record<string, Set<string>> = {};
  packages.forEach((pkg) => {
    const name = pkg.packageFile;
    packageFileNames.push(name);
    extractedPackages[name] = pkg;
    extractedDeps[name] = [];
  });

  // Construct package-specific prop scopes
  // and merge them in reverse order,
  // which allows inheritance/overriding.
  packageFileNames.forEach((name) => {
    registryUrls[name] = new Set();
    const propsHierarchy: Record<string, MavenProp>[] = [];
    const visitedPackages: Set<string> = new Set();
    let pkg: MavenInterimPackageFile | null = extractedPackages[name];
    while (pkg) {
      propsHierarchy.unshift(pkg.mavenProps!);

      if (pkg.deps) {
        pkg.deps.forEach((dep) => {
          if (dep.registryUrls) {
            dep.registryUrls.forEach((url) => {
              registryUrls[name].add(url);
            });
          }
        });
      }

      if (pkg.parent && !visitedPackages.has(pkg.parent)) {
        visitedPackages.add(pkg.parent);
        pkg = extractedPackages[pkg.parent];
      } else {
        pkg = null;
      }
    }
    propsHierarchy.unshift({});
    extractedProps[name] = Object.assign.apply(null, propsHierarchy as any);
  });

  // Resolve registryUrls
  packageFileNames.forEach((name) => {
    const pkg = extractedPackages[name];
    pkg.deps.forEach((rawDep) => {
      const urlsSet = new Set([...rawDep.registryUrls!, ...registryUrls[name]]);
      rawDep.registryUrls = [...urlsSet];
    });
  });

  const rootDeps = new Set<string>();
  // Resolve placeholders
  packageFileNames.forEach((name) => {
    const pkg = extractedPackages[name];
    pkg.deps.forEach((rawDep) => {
      const dep = applyProps(rawDep, name, extractedProps[name]);
      if (dep.depType === 'parent') {
        const parentPkg = extractedPackages[pkg.parent!];
        if (parentPkg && !parentPkg.parent) {
          rootDeps.add(dep.depName!);
        }
      }
      const sourceName = dep.propSource ?? name;
      extractedDeps[sourceName].push(dep);
    });
  });

  const packageFiles = packageFileNames.map((packageFile) => {
    const pkg = extractedPackages[packageFile];
    const deps = extractedDeps[packageFile];
    for (const dep of deps) {
      if (rootDeps.has(dep.depName!)) {
        dep.depType = 'parent-root';
      }
    }
    return { ...pkg, deps };
  });

  return packageFiles;
}

function cleanResult(packageFiles: MavenInterimPackageFile[]): PackageFile[] {
  packageFiles.forEach((packageFile) => {
    delete packageFile.mavenProps;
    delete packageFile.parent;
    packageFile.deps.forEach((dep) => {
      delete dep.propSource;
    });
  });
  return packageFiles;
}

export async function extractAllPackageFiles(
  _config: ExtractConfig,
  packageFiles: string[]
): Promise<PackageFile[]> {
  const packages: PackageFile[] = [];
  const additionalRegistryUrls: string[] = [];

  for (const packageFile of packageFiles) {
    const content = await readLocalFile(packageFile, 'utf8');
    if (!content) {
      logger.debug({ packageFile }, 'packageFile has no content');
      continue;
    }
    if (packageFile.endsWith('settings.xml')) {
      const registries = extractRegistries(content);
      if (registries) {
        logger.debug(
          { registries, packageFile },
          'Found registryUrls in settings.xml'
        );
        additionalRegistryUrls.push(...registries);
      }
    } else {
      const pkg = extractPackage(content, packageFile);
      if (pkg) {
        packages.push(pkg);
      } else {
        logger.trace({ packageFile }, 'can not read dependencies');
      }
    }
  }
  if (additionalRegistryUrls) {
    for (const pkgFile of packages) {
      for (const dep of pkgFile.deps) {
        /* istanbul ignore else */
        if (dep.registryUrls) {
          dep.registryUrls.push(...additionalRegistryUrls);
        } else {
          dep.registryUrls = [...additionalRegistryUrls];
        }
      }
    }
  }
  return cleanResult(resolveParents(packages));
}
