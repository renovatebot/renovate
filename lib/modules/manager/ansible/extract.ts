import { logger } from '../../../logger';
import { newlineRegex, regEx } from '../../../util/regex';
import * as dockerVersioning from '../../versioning/docker';
import { getDep } from '../dockerfile/extract';
import type {
  ExtractConfig,
  PackageDependency,
  PackageFileContent,
} from '../types';

export function extractPackageFile(
  content: string,
  packageFile: string,
  config: ExtractConfig
): PackageFileContent | null {
  logger.trace(`ansible.extractPackageFile(${packageFile})`);
  let deps: PackageDependency[] = [];
  const re = regEx(/^\s*image:\s*'?"?([^\s'"]+)'?"?\s*$/);
  for (const line of content.split(newlineRegex)) {
    const match = re.exec(line);
    if (match) {
      const currentFrom = match[1];
      const dep = getDep(currentFrom, true, config.registryAliases);
      logger.debug(
        {
          depName: dep.depName,
          currentValue: dep.currentValue,
          currentDigest: dep.currentDigest,
        },
        'Docker image inside ansible'
      );
      dep.versioning = dockerVersioning.id;
      deps.push(dep);
    }
  }
  deps = deps.filter((dep) => !dep.currentValue?.includes('${'));
  if (!deps.length) {
    return null;
  }
  return { deps };
}
