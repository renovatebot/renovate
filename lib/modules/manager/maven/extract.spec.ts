import { Fixtures } from '../../../../test/fixtures';
import { extractPackage, extractRegistries } from './extract';

const minimumContent = Fixtures.get(`minimum.pom.xml`);
const simpleContent = Fixtures.get(`simple.pom.xml`);

const mirrorSettingsContent = Fixtures.get(`mirror.settings.xml`);
const profileSettingsContent = Fixtures.get(`profile.settings.xml`);
const complexSettingsContent = Fixtures.get(`complex.settings.xml`);

describe('modules/manager/maven/extract', () => {
  describe('extractDependencies', () => {
    it('returns null for invalid XML', () => {
      expect(extractPackage('')).toBeNull();
      expect(extractPackage('invalid xml content')).toBeNull();
      expect(extractPackage('<foobar></foobar>')).toBeNull();
      expect(extractPackage('<project></project>')).toBeNull();
    });

    it('extract dependencies from any XML position', () => {
      const res = extractPackage(simpleContent);
      expect(res).toMatchSnapshot({
        deps: [
          {
            depName: 'org.example:parent',
            currentValue: '42',
            depType: 'parent',
          },
          {
            depName: 'org.example:foo',
            currentValue: '0.0.1',
            depType: 'compile',
          },
          {
            depName: 'org.example:bar',
            currentValue: '1.0.0',
            depType: 'compile',
          },
          {
            depName: 'org.apache.maven.plugins:maven-release-plugin',
            currentValue: '2.4.2',
            depType: 'build',
          },
          {
            depName: 'org.apache.maven.scm:maven-scm-provider-gitexe',
            currentValue: '1.8.1',
            depType: 'build',
          },
          {
            depName: 'org.example:extension-artefact',
            currentValue: '1.0',
            depType: 'build',
          },
          {
            depName: 'org.example:${artifact-id-placeholder}',
            currentValue: '0.0.1',
            depType: 'compile',
          },
          {
            depName: '${group-id-placeholder}:baz',
            currentValue: '0.0.1',
            depType: 'compile',
          },
          {
            depName: '${quuxGroup}:${quuxId}',
            currentValue: '${quuxVersion}',
            depType: 'compile',
          },
          {
            depName: '${quuxGroup}:${quuxId}-test',
            currentValue: '${quuxVersion}',
            depType: 'compile',
          },
          {
            depName: 'org.example:quuz',
            currentValue: '1.2.3',
            depType: 'test',
          },
          {
            depName: 'org.example:quuuz',
            currentValue: "it's not a version",
            depType: 'compile',
          },
          {
            depName: 'org.example:hard-range',
            currentValue: '[1.0.0]',
            depType: 'compile',
          },
          {
            depName: 'org.example:optional',
            currentValue: '1.0.0',
            depType: 'optional',
          },
          {
            depName: 'org.example:relocation-artifact',
            currentValue: '1.0',
          },
          {
            depName: 'org.example:profile-artifact',
            currentValue: '${profile-placeholder}',
            depType: 'compile',
          },
          {
            depName: 'org.example:profile-build-artefact',
            currentValue: '2.17',
            depType: 'build',
          },
          {
            depName: 'org.apache.maven.plugins:maven-checkstyle-plugin',
            currentValue: '2.17',
            depType: 'build',
          },
        ],
        mavenProps: {
          quuxGroup: {
            packageFile: null,
            val: 'org.example',
          },
          quuxId: {
            packageFile: null,
            val: 'quux',
          },
          quuxVersion: {
            packageFile: null,
            val: '1.2.3.4',
          },
        },
        packageFile: null,
      });
    });

    it('tries minimum manifests', () => {
      const res = extractPackage(minimumContent);
      expect(res).toEqual({
        datasource: 'maven',
        deps: [],
        mavenProps: {},
        packageFile: null,
        packageFileVersion: '1',
      });
    });
  });

  describe('extractRegistries', () => {
    it('returns null for invalid XML', () => {
      expect(extractRegistries('')).toBeEmptyArray();
      expect(extractRegistries('invalid xml content')).toBeEmptyArray();
      expect(extractRegistries('<foobar></foobar>')).toBeEmptyArray();
      expect(extractRegistries('<settings></settings>')).toBeEmptyArray();
    });

    it('extract registries from a simple mirror settings file', () => {
      const res = extractRegistries(mirrorSettingsContent);
      expect(res).toStrictEqual([
        'https://artifactory.company.com/artifactory/my-maven-repo',
      ]);
    });

    it('extract registries from a simple profile settings file', () => {
      const res = extractRegistries(profileSettingsContent);
      expect(res).toStrictEqual([
        'https://repo.adobe.com/nexus/content/groups/public',
      ]);
    });

    it('extract registries from a complex profile settings file', () => {
      const res = extractRegistries(complexSettingsContent);
      expect(res).toStrictEqual([
        'https://artifactory.company.com/artifactory/my-maven-repo',
        'https://repo.adobe.com/nexus/content/groups/public',
        'https://repo.adobe.com/v2/nexus/content/groups/public',
        'https://repo.adobe.com/v3/nexus/content/groups/public',
        'https://repo.adobe.com/v4/nexus/content/groups/public',
      ]);
    });
  });
});
