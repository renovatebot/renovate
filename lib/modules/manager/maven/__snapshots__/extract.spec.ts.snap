// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/manager/maven/extract extractDependencies extract dependencies from any XML position 1`] = `
{
  "datasource": "maven",
  "deps": [
    {
      "currentValue": "42",
      "datasource": "maven",
      "depName": "org.example:parent",
      "depType": "parent",
      "fileReplacePosition": 186,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "0.0.1",
      "datasource": "maven",
      "depName": "org.example:foo",
      "depType": "compile",
      "fileReplacePosition": 905,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "1.0.0",
      "datasource": "maven",
      "depName": "org.example:bar",
      "depType": "compile",
      "fileReplacePosition": 1093,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "2.4.2",
      "datasource": "maven",
      "depName": "org.apache.maven.plugins:maven-release-plugin",
      "depType": "build",
      "fileReplacePosition": 1347,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "1.8.1",
      "datasource": "maven",
      "depName": "org.apache.maven.scm:maven-scm-provider-gitexe",
      "depType": "build",
      "fileReplacePosition": 1545,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "1.0",
      "datasource": "maven",
      "depName": "org.example:extension-artefact",
      "depType": "build",
      "fileReplacePosition": 2276,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "0.0.1",
      "datasource": "maven",
      "depName": "org.example:\${artifact-id-placeholder}",
      "depType": "compile",
      "fileReplacePosition": 2484,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "0.0.1",
      "datasource": "maven",
      "depName": "\${group-id-placeholder}:baz",
      "depType": "compile",
      "fileReplacePosition": 2634,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "\${quuxVersion}",
      "datasource": "maven",
      "depName": "\${quuxGroup}:\${quuxId}",
      "depType": "compile",
      "fileReplacePosition": 2779,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "\${quuxVersion}",
      "datasource": "maven",
      "depName": "\${quuxGroup}:\${quuxId}-test",
      "depType": "compile",
      "fileReplacePosition": 2938,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "1.2.3",
      "datasource": "maven",
      "depName": "org.example:quuz",
      "depType": "test",
      "fileReplacePosition": 3086,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "it's not a version",
      "datasource": "maven",
      "depName": "org.example:quuuz",
      "depType": "compile",
      "fileReplacePosition": 3252,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "[1.0.0]",
      "datasource": "maven",
      "depName": "org.example:hard-range",
      "depType": "compile",
      "fileReplacePosition": 3410,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "1.0.0",
      "datasource": "maven",
      "depName": "org.example:optional",
      "depType": "optional",
      "fileReplacePosition": 3555,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "1.0",
      "datasource": "maven",
      "depName": "org.example:relocation-artifact",
      "fileReplacePosition": 3787,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "\${profile-placeholder}",
      "datasource": "maven",
      "depName": "org.example:profile-artifact",
      "depType": "compile",
      "fileReplacePosition": 4119,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "2.17",
      "datasource": "maven",
      "depName": "org.example:profile-build-artefact",
      "depType": "build",
      "fileReplacePosition": 4375,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
    {
      "currentValue": "2.17",
      "datasource": "maven",
      "depName": "org.apache.maven.plugins:maven-checkstyle-plugin",
      "depType": "build",
      "fileReplacePosition": 4769,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2",
        "https://maven.atlassian.com/content/repositories/atlassian-public/",
      ],
    },
  ],
  "mavenProps": {
    "quuxGroup": {
      "fileReplacePosition": 631,
      "packageFile": null,
      "val": "org.example",
    },
    "quuxId": {
      "fileReplacePosition": 667,
      "packageFile": null,
      "val": "quux",
    },
    "quuxVersion": {
      "fileReplacePosition": 698,
      "packageFile": null,
      "val": "1.2.3.4",
    },
  },
  "packageFile": null,
  "packageFileVersion": "0.0.1",
}
`;
